README.md

```diff
@@ -0,0 +1,121 @@
+# Kubectl Wrapper Modules For Ansible
+
+The recommended approach for declarative management of Kubernetes config is
+`kubectl apply`, which contains sophisticated logic for performing a three way
+merge between an an objects incoming config, last applied config, and current
+state.  This logic however is implemented client side, and thus not available
+to callers of the API (including the current kubernetes module), and non-Go
+languages. In the future this logic may move to the API, however this is not
+expected near term.
+
+However kubectl is intended to be treated as a type of API, as such in the interim we can leverage it's logic by wrapping in Ansible modules.
+
+This prototype is specific to generically managing actual Kubernetes config, it
+does not wrap Kubernetes yaml in Ansible yaml, or expose individual modules per
+individual Kubernetes type.
+
+This prototype aims to make applying Kubernetes config via Ansible extremely
+easy, including on-the-fly templating and simple yaml edits, file based or
+inline config, and patching.
+
+`kubectl delete` is represented by state: absent on the kubectl module itself.
+
+`kubectl get` is represented by a kubectl_facts module.
+
+## Assumptions
+
+  * kubectl apply will be modified to accurately [report if changes had to be applied](https://github.com/kubernetes/kubernetes/pull/52647).
+
+# Target Ansible
+
+```yaml
+- kubectl:
+
+    # State present implies kubectl apply, and absent implies delete. Many options
+    # below would not be relevant for delete.
+    state: present
+
+    kubeconfig:
+      # Kubeconfig file on the remote host:
+      file: /etc/origin/master/admin.kubeconfig
+      # Alternatively:
+      inline: "{{ lookup('file', 'local_on_control_host.kubeconfig') }}"
+
+    # Optional, omit if your configuration specifies a namespace, or you're operating on
+    # cluster level objects.
+    namespace: default
+
+    # You can specify your kubernetes config inline if desired. The | is important here:
+    inline: |
+      kind: "Namespace"
+      apiVersion: v1
+      metadata:
+        name: testnamespace
+
+    # Alternatively you can use files that live with your role/playbook. This would be
+    # mutually exclusive with inline data.
+    # Each file is copied over to the remote host, potentially edited or patched, and
+    # then submitted with apply.
+    # Specifying multiple files implies one execution of kubectl apply. If you want multiple
+    # you can use separate tasks or a with_items loop.
+    files:
+    - src: configmap.json
+    - src: subdir/
+    - src: service.yml
+      # Support small edits to the yaml on the fly, without going to a full template. Operates
+      # on the temporary copy of the file sent to the remote host.
+      yedit:
```

Comment:
I think it's probably better to call out yedit as a separate, valuable module.

Reply:
Could it still be embedded for use here just like templating and patching? If not, users now need to think about writing files out to specific locations on the remote host and cleaning them up, whereas I'd like to see this just handle all of that for you. User likely doesn't care about having to place files on remote server, then apply them in separate ansible tasks, then potentially clean them up.